urlcdab7d8364_jrwalsh_CycTools_tgz-pJ8-edu_iastate_cyctools_MainJ8/src/edu/iastate/cyctools/externalSourceCode/DiffMatchPatch.java

```diff
@@ -0,0 +1,2535 @@
+          Math.min(text.length(), patch.start2 + patch.length1 + padding));
+    }
+    // Add one chunk for good luck.
+    padding += Patch_Margin;
+
+    // Add the prefix.
+    String prefix = text.substring(Math.max(0, patch.start2 - padding),
+        patch.start2);
+    if (prefix.length() != 0) {
+      patch.diffs.addFirst(new Diff(Operation.EQUAL, prefix));
+    }
+    // Add the suffix.
+    String suffix = text.substring(patch.start2 + patch.length1,
+        Math.min(text.length(), patch.start2 + patch.length1 + padding));
+    if (suffix.length() != 0) {
+      patch.diffs.addLast(new Diff(Operation.EQUAL, suffix));
+    }
+
+    // Roll back the start points.
+    patch.start1 -= prefix.length();
+    patch.start2 -= prefix.length();
+    // Extend the lengths.
+    patch.length1 += prefix.length() + suffix.length();
+    patch.length2 += prefix.length() + suffix.length();
+  }
+
+  /**
+   * Compute a list of patches to turn text1 into text2.
+   * A set of diffs will be computed.
+   * @param text1 Old text.
+   * @param text2 New text.
+   * @return LinkedList of Patch objects.
+   */
+  public LinkedList<Patch> patch_make(String text1, String text2) {
+    if (text1 == null || text2 == null) {
+      throw new IllegalArgumentException("Null inputs. (patch_make)");
+    }
+    // No diffs provided, compute our own.
+    LinkedList<Diff> diffs = diff_main(text1, text2, true);
+    if (diffs.size() > 2) {
+      diff_cleanupSemantic(diffs);
+      diff_cleanupEfficiency(diffs);
+    }
+    return patch_make(text1, diffs);
+  }
+
+  /**
+   * Compute a list of patches to turn text1 into text2.
+   * text1 will be derived from the provided diffs.
+   * @param diffs Array of Diff objects for text1 to text2.
+   * @return LinkedList of Patch objects.
+   */
+  public LinkedList<Patch> patch_make(LinkedList<Diff> diffs) {
+    if (diffs == null) {
+      throw new IllegalArgumentException("Null inputs. (patch_make)");
+    }
+    // No origin string provided, compute our own.
+    String text1 = diff_text1(diffs);
+    return patch_make(text1, diffs);
+  }
+
+  /**
+   * Compute a list of patches to turn text1 into text2.
+   * text2 is ignored, diffs are the delta between text1 and text2.
+   * @param text1 Old text
+   * @param text2 Ignored.
+   * @param diffs Array of Diff objects for text1 to text2.
+   * @return LinkedList of Patch objects.
+   * @deprecated Prefer patch_make(String text1, LinkedList<Diff> diffs).
+   */
+  public LinkedList<Patch> patch_make(String text1, String text2,
+      LinkedList<Diff> diffs) {
+    return patch_make(text1, diffs);
+  }
+
+  /**
+   * Compute a list of patches to turn text1 into text2.
+   * text2 is not provided, diffs are the delta between text1 and text2.
+   * @param text1 Old text.
+   * @param diffs Array of Diff objects for text1 to text2.
+   * @return LinkedList of Patch objects.
+   */
+  public LinkedList<Patch> patch_make(String text1, LinkedList<Diff> diffs) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.